### DISCUSSION TITLE
Unable to find a great way of guaranteeing a connection isn't expired

### DISCUSSION THREAD

**Discussion by justin-richert** [psycopg/psycopg#850] [Ideas]
Hello! Thanks for such a great python library!

I'm running into an issue that I thought I might get your thoughts on, either to perhaps spark an idea on an improvement that could be made to the library to solve this or to prompt a correction in how I'm attempting to solve my problem.

**Problem**: In a Kubernetes environment where the pgbouncer instances we are connecting to can be removed at just about any time for scaling or node draining purposes, I need a reliable way to ensure that I've properly released connections from a pool before a pod goes away. Because there is no indicator to the application that pgbouncer is _about_ to shut down, we've implemented a graceful shutdown strategy for our pgbouncer pods paired with a connection recycle time that is just a little bit shorter than the graceful shutdown window.

_However_, because Psycopg is not checking for connection expiration until said connection is being put back into the pool after use*, there are still failure scenarios where an expired connection sits around for some amount of time before actually being picked up to be used. So the scenario would be, a connection is opened, given a 1 minute expiration period, used, put back in the pool, goes unused for 2 minutes, is attempted to be used again, _then_ checked for expiration and discarded. The "attempted to be used again" is where we're having some trouble, as the connection may no longer be good at that time.

**Attempted Solution**: Use the `check` argument to supply a function that checks for connection expiration when grabbing connections. This _mostly_ works. A couple of issues, though, the latter of which lead me to open this issue:
- We have to look at a "private" attribute on the connection (`conn._expire_at`) in order to do this. Not a huge deal, but probably not something you guys are considering part of the maintained and guaranteed interface.
- In scenarios where many connections expire around the same time, this coupled with the retry backoff (which is only detrimental and not actually helpful in our situation), it is pretty easy to cause `PoolTimeout` errors, especially if our applications are configured with low but reasonable timeout settings in a production environment (something like 2 seconds, mainly to account for these situations to try to reduce these occurrences). This can be caused _in part_ by a pool that isn't rightsized, but there's all sorts of reasons why a dev may allow for a large pool size but only need it in rare instances. Although I certainly consider that part of our solution, I don't like that our team is having to make considerations like, "should this pool size be reduced from 20 to 10 to avoid this?". Even if 99% of the time 10 was enough, the dev _allowing_ for 20 seems like it would be reasonable, and making them work this hard to right size the pool due to some of this implementation detail stuff feels like a bit of a distraction from the actual work they're engaging in.

**Assumptions**:
- Connections are only checked for expiration purposes when putting them back in the pool (marked with an asterisk above). This isn't a blind assumption but more of a conclusion drawn from reading the code, I just want to acknowledge here that this could be a misunderstanding on my part. However, certain tests have appeared to confirm this behavior.

Thank you so much for reading and for your consideration. Please don't hesitate to let me know if there are any questions or any clarifications I need to make.

**Comment by justin-richert** [comment]
I've realized I left out what was going to be a proposed solution. Not sure it will be helpful, but I'll add it anyway just in case.

I was hoping it might make sense to allow for just creating a new "replacement" connection and returning it to the code asking for a connection either after the first check function failure or at least a low number of failures to find a non-expired connection. We tried implementing this ourselves as a solution, but this ended up in us having to effectively copy a lot of the new pool connection logic from Psycopg into our own code (to do things like setting the `_expire_at` variable, attaching it to the pool, etc.), and that was just a non-starter for me. That code could obviously easily change without us knowing and then we might be creating connections for the pool that don't look like those created from within the library.

**Comment by dvarrazzo** [comment]
Issue converted to a discussion.

I am not against someone providing a solution to this problem, providing it's generic enough and that it is well tested. But I am personally not available to provide a solution for free.

If you would like to propose a paid collaboration, feel free to reach out privately.

**Reply by justin-richert** [reply]
Ok, great. Thanks for the reply / info @dvarrazzo! I will get back to you as quickly as I can on this.
